File: app/src/main/java/com/maliotis/petros/weather/ExpDayAdapter.java
package com.maliotis.petros.weather;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseExpandableListAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import com.maliotis.petros.weather.weather.Day;
import com.maliotis.petros.weather.weather.Hour;


public class ExpDayAdapter extends BaseExpandableListAdapter {
    private Context mContext;
    private Day[] mDays;
    private Hour[][] mHours;

    public ExpDayAdapter(Context context, Day[] days, Hour[][] hours){
        mContext = context;
        mDays = days;
        mHours = hours;
    }

    @Override
    public int getGroupCount() {
        return mDays.length;
    }

    @Override
    public int getChildrenCount(int groupPosition) {
        return 8;
    }

    @Override
    public Object getGroup(int groupPosition) {
        return mDays[groupPosition];
    }

    @Override
    public Object getChild(int groupPosition, int childPosition) {
        return mHours[groupPosition][childPosition];
    }

    @Override
    public long getGroupId(int groupPosition) {
        return 0;
        //return groupPosition;
    }

    @Override
    public long getChildId(int groupPosition, int childPosition) {
        return 0;
        //return childPosition;
    }

    @Override
    public boolean hasStableIds() {
        return false;
    }

    @Override
    public View getGroupView(int groupPosition, boolean isExpanded, View convertView, ViewGroup parent) {
        ViewHolderG viewHolderG;

        if(convertView == null){
            //brand new

            convertView = LayoutInflater.from(mContext).inflate(R.layout.daily_list_item,null);
            viewHolderG = new ViewHolderG();
            viewHolderG.iconImageView = (ImageView) convertView.findViewById(R.id.iconImageView);
            viewHolderG.temperatureLabel = (TextView) convertView.findViewById(R.id.temperatureLabel);
            viewHolderG.dayLabel = (TextView) convertView.findViewById(R.id.dayNameLabel);

            convertView.setTag(viewHolderG);
        }
        else{
            viewHolderG = (ViewHolderG) convertView.getTag();
        }

        Day day = mDays[groupPosition];
        viewHolderG.iconImageView.setImageResource(day.getIconId());
        String temp = day.getTemperatureMax()+"";
        viewHolderG.temperatureLabel.setText(temp);
        if(groupPosition == 0){
            viewHolderG.dayLabel.setText(R.string.Today);
        }
        else {
            viewHolderG.dayLabel.setText(day.getDayOfTheWeek());
        }

        return convertView;
    }

    @Override
    public View getChildView(int groupPosition, int childPosition, boolean isLastChild, View convertView, ViewGroup parent) {
        ViewHolderChild holderChild;

        if(convertView == null){
            //brand new
            convertView = LayoutInflater.from(mContext).inflate(R.layout.hourly_list_item,null);
            holderChild = new ViewHolderChild();
            holderChild.mTimeLabel = (TextView) convertView.findViewById(R.id.timeLabel);
            holderChild.mSummaryLabel = (TextView) convertView.findViewById(R.id.summaryLabel);
            holderChild.mTemperatureLabel = (TextView) convertView.findViewById(R.id.temperatureLabel);
            holderChild.mIconImageView = (ImageView) convertView.findViewById(R.id.iconImageView);
            holderChild.mHourDay = (TextView) convertView.findViewById(R.id.hourDayLabel);
            holderChild.mDegree = (TextView) convertView.findViewById(R.id.degreeLabel);
            convertView.setTag(holderChild);

        }
        else {
            holderChild = (ViewHolderChild) convertView.getTag();
        }

        Hour hours = mHours[groupPosition][childPosition];
        holderChild.mTimeLabel.setText(hours.getHour());
        holderChild.mSummaryLabel.setText(hours.getSummary());
        holderChild.mTemperatureLabel.setText(hours.getTemperature()+"");
        holderChild.mIconImageView.setImageResource(hours.getIconId());
        holderChild.mHourDay.setText("");
        holderChild.mDegree.setText("o");


        return convertView;
    }

    @Override
    public boolean isChildSelectable(int groupPosition, int childPosition) {
        return false;
    }

    private static class ViewHolderG{
        ImageView iconImageView;
        TextView temperatureLabel;
        TextView dayLabel;
    }

    private static class ViewHolderChild{

        TextView mTimeLabel;
        TextView mSummaryLabel;
        TextView mTemperatureLabel;
        ImageView mIconImageView;
        TextView mHourDay;
        TextView mDegree;
    }
}

File: app/src/main/java/com/maliotis/petros/weather/ExpDayAdapter.kt
package com.maliotis.petros.weather

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseExpandableListAdapter
import android.widget.ImageView
import android.widget.TextView
import com.maliotis.petros.weather.weather.Day
import com.maliotis.petros.weather.weather.Hour

class ExpDayAdapter(private val mContext: Context, private val mDays: Array<Day?>, private val mHours: Array<Array<Hour?>>) : BaseExpandableListAdapter() {
    override fun getGroupCount(): Int {
        return mDays.size - 1
    }

    override fun getChildrenCount(groupPosition: Int): Int {
        return 8
    }

    override fun getGroup(groupPosition: Int): Day? {
        return mDays[groupPosition]
    }

    override fun getChild(groupPosition: Int, childPosition: Int): Any? {
        return mHours[groupPosition][childPosition]
    }

    override fun getGroupId(groupPosition: Int): Long {
        return 0
        //return groupPosition;
    }

    override fun getChildId(groupPosition: Int, childPosition: Int): Long {
        return 0
        //return childPosition;
    }

    override fun hasStableIds(): Boolean {
        return false
    }

    override fun getGroupView(groupPosition: Int, isExpanded: Boolean, convView: View?, parent: ViewGroup): View {
        var convertView = convView
        val viewHolderG: ViewHolderG
        if (convertView == null) {
            //brand new
            convertView = LayoutInflater.from(mContext).inflate(R.layout.daily_list_item, null)
            viewHolderG = ViewHolderG()
            viewHolderG.iconImageView = convertView.findViewById<View>(R.id.iconImageView) as ImageView
            viewHolderG.temperatureLabel = convertView.findViewById<View>(R.id.temperatureLabel) as TextView
            viewHolderG.minTempLabel = convertView.findViewById(R.id.minTempLabel)
            viewHolderG.dayLabel = convertView.findViewById<View>(R.id.dayNameLabel) as TextView
            convertView.tag = viewHolderG
        } else {
            viewHolderG = convertView.tag as ViewHolderG
        }
        val day = mDays[groupPosition]
        day?.iconId?.run {
            viewHolderG.iconImageView!!.setImageResource(this)
        }

        val temp = day?.temperatureMax.toString() + "°C"
        val minTemp = day?.temperatureMin.toString() + "°C"
        viewHolderG.temperatureLabel!!.text = temp
        viewHolderG.minTempLabel!!.text = minTemp
        if (groupPosition == 0) {
            viewHolderG.dayLabel!!.setText(R.string.Today)
        } else {
            viewHolderG.dayLabel!!.text = day?.dayOfTheWeek
        }
        return convertView!!
    }

    override fun getChildView(groupPosition: Int, childPosition: Int, isLastChild: Boolean, convView: View?, parent: ViewGroup): View {
        var convertView = convView
        val holderChild: ViewHolderChild
        if (convertView == null) {
            //brand new
            convertView = LayoutInflater.from(mContext).inflate(R.layout.hourly_list_item, null)
            holderChild = ViewHolderChild()
            holderChild.mTimeLabel = convertView.findViewById<View>(R.id.timeLabel) as TextView
            holderChild.mSummaryLabel = convertView.findViewById<View>(R.id.summaryLabel) as TextView
            holderChild.mTemperatureLabel = convertView.findViewById<View>(R.id.temperatureLabel) as TextView
            holderChild.mIconImageView = convertView.findViewById<View>(R.id.iconImageView) as ImageView
            holderChild.mHourDay = convertView.findViewById<View>(R.id.hourDayLabel) as TextView
            holderChild.mDegree = convertView.findViewById<View>(R.id.degreeLabel) as TextView
            convertView.tag = holderChild
        } else {
            holderChild = convertView.tag as ViewHolderChild
        }
        val hours = mHours[groupPosition][childPosition]
        holderChild.mTimeLabel!!.text = hours?.hour
        holderChild.mSummaryLabel!!.text = hours?.summary
        holderChild.mTemperatureLabel!!.text = hours?.temperature.toString() + ""
        hours?.iconId?.run {
            holderChild.mIconImageView!!.setImageResource(this)
        }
        holderChild.mHourDay!!.text = ""
        holderChild.mDegree!!.text = "o"
        return convertView!!
    }

    override fun isChildSelectable(groupPosition: Int, childPosition: Int): Boolean {
        return false
    }

    private class ViewHolderG {
        var iconImageView: ImageView? = null
        var temperatureLabel: TextView? = null
        var dayLabel: TextView? = null
        var minTempLabel: TextView? = null
    }

    private class ViewHolderChild {
        var mTimeLabel: TextView? = null
        var mSummaryLabel: TextView? = null
        var mTemperatureLabel: TextView? = null
        var mIconImageView: ImageView? = null
        var mHourDay: TextView? = null
        var mDegree: TextView? = null
    }

}